peps/pep-0667.rst

```diff
@@ -169,47 +331,183 @@ The following functions should be used instead::
+        exec('print(a)', locals=ns)  # 0
+    f()
+
+You can even reliably change the variables in the local scope by explicitly using
+``frame.f_locals``, which was not possible before (even using ``ctypes`` to
+invoke ``PyFrame_LocalsToFast`` was subject to the state inconsistency problems
+discussed elsewhere in this PEP)::
+
+    def f():
+        a = None
+        exec('a = 0', locals=sys._getframe().f_locals)
+        print(a)  # 0
+    f()
+
+The behavior of ``exec()`` and ``eval()`` for module and class scopes (including
+nested invocations) is not changed, as the behaviour of ``locals()`` in those
+scopes is not changing.
+
+Impact on other code execution APIs in the standard library
+'''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
+
+``pdb`` and ``bdb`` use the ``frame.f_locals`` API, and hence will be able to
+reliably update local variables even in optimized frames. Implementing this
+PEP will resolve several longstanding bugs in these modules relating to threads,
+generators, coroutines, and other mechanisms that allow concurrent code execution
+while the debugger is active.
+
+Other code execution APIs in the standard library (such as the ``code`` module)
+do not implicitly access ``locals()`` *or* ``frame.f_locals``, but the behaviour
+of explicitly passing these namespaces will change as described in the rest of
+this PEP (passing ``locals()`` in optimized scopes will no longer implicitly
+share the code execution namespace across calls, passing ``frame.f_locals``
+in optimized scopes will allow reliable modification of local variables and
+nonlocal cell references).
+
+C API compatibility
+-------------------
 
 PyEval_GetLocals
 ''''''''''''''''
 
-Because ``PyEval_GetLocals()`` returns a borrowed reference, it requires
-the proxy mapping to be cached on the frame, extending its lifetime and
-creating a cycle. ``PyEval_GetFrameLocals()`` should be used instead.
+``PyEval_GetLocals()`` has never historically distinguished between whether it was
+emulating ``locals()`` or ``sys._getframe().f_locals`` at the Python level, as they all
+returned references to the same shared cache of the local variable bindings.
+
+With this PEP, ``locals()`` changes to return independent snapshots on each call for
+optimized frames, and ``frame.f_locals`` (along with ``PyFrame_GetLocals``) changes to
+return new write-through proxy instances.
+
+Because ``PyEval_GetLocals()`` returns a borrowed reference, it isn't possible to update
+its semantics to align with either of those alternatives, leaving it as the only remaining
+API that requires a shared cache dictionary stored on the frame object.
+
+While this technically leaves the semantics of the function unchanged, it no longer allows
+extra dict entries to be made visible to users of the other APIs, as those APIs are no longer
+accessing the same underlying cache dictionary.
+
+Accordingly, the function will be marked as soft-deprecated in Python 3.13, and subsequently
```

Comment:
Makes sense, I'll reword this bit.

Reply:
Moved to the specification section, changed to "deprecated in the documentation" and "emit DeprecationWarning" to avoid the soft vs hard deprecation jargon.